### ISSUE TITLE
Adding more modules like `metrics` and `losses`

### ISSUE THREAD

**Issue by AakashKumarNain** [google/jax#3721]
`JAX` is a pretty cool frameworks and it has the potential to be the next big thing in the ML world. Though it is in very early stage, it has already gained a huge momentum and people have come up with their own frameworks built on top of JAX.

### What is this about?
Even though JAX is quite popular, there are some low level functionalities that are missing from it, for examples `metrics` and `losses` modules. 

### Why does it matter?
One can argue that these requirements don't fit in the core library because people can implement these on their own as they are pretty easy. But that implication has its own problems. A few points in this regard:
1. There are going to be many many frameworks that will be built on top of JAX and each of them would be implementing their own version of these functionalities. Although from the perspective of a core framework that sounds okay but if we look from the perspective of an `ecosystem`, it is pretty bad. More the number of implementations, more the fragmentation in the ecosystem. Hence, these things should be built inside the core and any new framework should directly these functionalities from the core only. Not only it makes the ecosystem more cohesive, it helps the contributors and maintainers to track and solve the potential bugs.
2. This also reduces the workload for anyone who is trying to build things on top of JAX as it doesn't need to be implemented from scratch.

I know that this will take some time and efforts to implement it but if the maintainers are okay with this, I am willing to contribute to these. 

cc: @mattjj

**Comment by shoyer** [comment]
We have generally been open to adding functions into jax.nn for commonly used and/or numerically tricky functions. This could well include some functions for computing losses or metrics, e.g., perhaps for softmax cross entropy.

My personal experience is that the value of generic abstractions for simple functions like mean squared error or hinge loss (i.e., most of what you can find in keras.losses) is marginal if not negative for research code:

- The idiomatic implementation of these functions in terms of NumPy is often ~1 line of code, which is hardly a maintenance burden to fork into each project that uses them. 
- Obscuring the implementation of metrics makes it harder for users to add their own or make modifications.
- By putting losses/metrics in a shared library, there’s much more code to keep in sync. If our perspective on the optimal design for losses/metrics changes, now all client code needs to be updated to use the new signature or functions. This is not entirely hypothetical — in TF, the signature for various loss functions  was updated a handful of times.

So here’s my (non-official) outlook on this question:

- I think a generic “metrics” or “losses” library would be a better suited for a downstream project with an opinionated design about how to structure ML models rather than in JAX proper.
- A pattern that I like better would be to provide forkable “recipes” for metrics/losses in our documentation, like those documented alongside itertools in the Python standard library.

**Comment by AakashKumarNain** [comment]
@shoyer yes, I agree that it surely increases the burden to maintain a flexible API and it is also hard to cover all the use cases within the API.

PS: Is there any other functionality in JAX we can contribute to? The [Contribution guide](https://github.com/google/jax/blob/master/CONTRIBUTING.md) or the README doesn't specify the areas where the maintainers can get some help from the community. Thanks again for looking into it

**Comment by cyprienc** [comment]
Sorry for reviving this - was there a `metrics` open-source library/module implemented in the end?
Thanks !
